crates/hyperdrive-math/src/lp.rs

```diff
@@ -4,9 +4,144 @@ use ethers::types::{I256, U256};
 use fixed_point::FixedPoint;
 use fixed_point_macros::{fixed, int256};
 
-use crate::{State, YieldSpace};
+use crate::{calculate_effective_share_reserves, State, YieldSpace};
 
 impl State {
+    // Calculates the lp_shares for a given contribution when adding liquidity.
+    pub fn calculate_add_liquidity(
+        &self,
+        current_block_timestamp: U256,
+        contribution: FixedPoint,
+        min_apr: FixedPoint,
+        max_apr: FixedPoint,
+        as_base: bool,
+    ) -> Result<FixedPoint, &str> {
+        // Ensure that the contribution is greater than or equal to the minimum
+        // transaction amount.
+        if contribution < self.minimum_transaction_amount() {
+            return Err("MinimumTransactionAmount");
+        }
+
+        // Enforce the slippage guard.
+        let apr = self.calculate_spot_rate();
+        if apr < min_apr || apr > max_apr {
+            return Err("InvalidApr");
+        }
+
+        // Get lp_total_supply for the lp_shares calculation.
+        let lp_total_supply = self.lp_total_supply();
+
+        // Get the starting_present_value.
+        let starting_present_value = self.calculate_present_value(current_block_timestamp);
+
+        // Get the ending_present_value.
+        let share_contribution = {
+            if as_base {
+                // Attempt a crude conversion from base to shares.
+                I256::try_from(contribution / self.vault_share_price()).unwrap()
+            } else {
+                I256::try_from(contribution).unwrap()
+            }
+        };
+        let new_state = self.update_liquidity(share_contribution);
+        let ending_present_value = new_state.calculate_present_value(current_block_timestamp);
+
+        // Ensure the present value didn't decreased after adding liquidity.
+        if ending_present_value < starting_present_value {
+            return Err("DecreasedPresentValueWhenAddingLiquidity");
+        }
+
+        // Calculate the lp_shares.
+        let lp_shares = (ending_present_value - starting_present_value)
+            .mul_div_down(lp_total_supply, starting_present_value);
+
+        // Ensure that enough lp_shares are minted so that they can be redeemed.
+        if lp_shares < self.minimum_transaction_amount() {
+            return Err("MinimumTransactionAmount");
+        }
+
+        return Ok(lp_shares);
+    }
+
+    // Gets the resulting state when updating liquidity.
+    pub fn update_liquidity(&self, share_reserves_delta: I256) -> State {
```

Comment:
I think a better name for this would be something like `get_state_after_liquidity_update`. To me `update_liquidity` implies that it would happen in-place.